design/crypto_primitives.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Different MaidSafe tools need different crypto primitives. E.g. Crust only uses encryption
+//! public/private keys, Routing uses encryption and signing public/private keys, etc.
+//! Currently in some cases safe_crypto is unusable cause it doesn't expose some internal data like
+//! private keys, etc.
+//! One way to solve this and stay super flexible is to have 1-to-1 mappings with lower level
+//! crypto types from rust_sodium, miscreant, etc.
+//! This document attempts to catch safe_crypto design that would work for all our our cases.
+//!
+//! `PublicKeys` and `SecretKeys` currently contain both signing and encryption keys. In some
+//! cases, we don't need signing key at all, e.g. in Crust. So this document proposes to split
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// no signing key
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    // No new() method, since public key should be constructed using `SecretEncryptKey`
+    // pub fn new() -> Self
+
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: &[u8]) -> Self
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(&self) -> &[u8]
```

Comment:
These set of functions i think should return an array instead of a slice as it allows for users to consume it as arrays rather than as heap allocated storage. With slices, the length is expressed only at runtime forcing the collection into a heap allocated container (like `Vec` for e.g.)

There should perhaps be a contant:
```rust
const PUBLIC_ENCRYPT_KEY_BYTES: usize = <n>; // different for mock and non-mock ofc;
                                             // also is `_` favoured for constants ?
...
pub fn raw(&self) -> [u8; PUBLIC_ENCRYPT_KEY_BYTES];
```
Now the consumption can be as:
```
let consume = enc_pk.raw();
```